# Introducción a Python para IA.

 <p xmlns:cc="http://creativecommons.org/ns#" xmlns:dct="http://purl.org/dc/terms/"><a property="dct:title" rel="cc:attributionURL" href="https://github.com/luiggix/intro_MeIA_2023">Introducción a Python para IA</a> by <span property="cc:attributionName">Luis Miguel de la Cruz Salas</span> is licensed under <a href="http://creativecommons.org/licenses/by-nc-sa/4.0/?ref=chooser-v1" target="_blank" rel="license noopener noreferrer" style="display:inline-block;">CC BY-NC-SA 4.0<img style="height:22px!important;margin-left:3px;vertical-align:text-bottom;" src="https://mirrors.creativecommons.org/presskit/icons/cc.svg?ref=chooser-v1"><img style="height:22px!important;margin-left:3px;vertical-align:text-bottom;" src="https://mirrors.creativecommons.org/presskit/icons/by.svg?ref=chooser-v1"><img style="height:22px!important;margin-left:3px;vertical-align:text-bottom;" src="https://mirrors.creativecommons.org/presskit/icons/nc.svg?ref=chooser-v1"><img style="height:22px!important;margin-left:3px;vertical-align:text-bottom;" src="https://mirrors.creativecommons.org/presskit/icons/sa.svg?ref=chooser-v1"></a></p> 

# Objetivos.

Hacer uso de los conceptos de etiquetas, palabras reservadas, expresiones, declaraciones, tipos y operadores, importación de bibliotecas (módulos), estructura de datos, control de flujo, entrada y salida estándar para implementar una fórmula matemática.

# El huevo cocido perfecto

¿Cuál es el tiempo ideal para cocinar un huevo?

<img src="../utils/fig/huevo_cocido.jpg"  style="width: 500px;"/>


Para lograr un **huevo cocido** suave, la clara debe haberse calentado el tiempo suficiente para coagular a una temperatura superior a $63^oC$, pero la yema no debe calentarse por encima de $70^oC$.

Para lograr un **huevo duro**, el centro de la yema debe de alcanzar los $70^oC$.

## Fórmula para calcular el tiempo de cocción.
La siguiente fórmula expresa el tiempo $t$, en segundos, que le toma a la yema alcanzar la temperatura $T_y$, en grados Celsius.

$$t = \dfrac{M^{2/3} c \rho^{1/3}}{K \pi^2 (4\pi/3)^{2/3}} \ln \left[ 0.76 \dfrac{T_o - T_w}{T_y - T_w}\right]$$

donde las propiedades son:

- $M$ masa; 
- $\rho$ densidad; 
- $c$ capacidad calorífica específica; 
- $K$ conductividad térmica; 
- $T_w$ es la temperatura de ebullición del agua;
- $T_o$ es la temperatura original del huevo antes de meterlo al agua;
- $T_y$ es la temperatura que debe alcanzar la yema.

---
## **<font color="DodgerBlue">Ejercicio 1. </font>**

<font color="DarkBlue">
Calcular el tiempo de cocción necesario para un huevo duro pequeño de $M$ = 47 g 
(la masa de un huevo grande es 67 g), para cuando la temperatura inicial 
del huevo es:
    
1. Temperatura ambiente: $T_o = 20^oC$.
2. Temperatura en el refrigerador: $T_o = 4^oC$.

**Datos**:
- $M$ = 47 g 
- $\rho$ = 1.038 g / cm$^3$
- $c$ = 3.7 J / g K
- $K$ = 5.4 $\times 10^{-3}$ W / cm K
- $T_w = 100^oC$
- $T_y = 70^oC$
</font>

---


### Paso 1. Numerador.
Calcular la primera parte de la fórmula: numerador = $M^{2/3} c \rho^{1/3}$. Definir los elementos de esta primera parte de la fórmula de tal manera que: $M$ sea un entero, $\rho$ y $c$ flotantes.

In [1]:
# Definimos las etiquetas necesarias
M   = 47    # Un entero
rho = 1.038 # Un flotante
c   = 3.7   # Otro flotante

# Imprimimos el contenido, el tipo y el id de cada objeto
print(M, type(M), id(M))
print(rho, type(rho), id(rho))
print(c, type(c), id(c))

47 <class 'int'> 140732205037800
1.038 <class 'float'> 2931489792880
3.7 <class 'float'> 2931489794288


Con los datos anteriores cálculamos el `numerador`:

In [2]:
numerador = M**(2/3) * c * rho**(1/3)
print(numerador, type(numerador), id(numerador))

48.790216719661984 <class 'float'> 2931489793488


<div class="alert alert-success">
<b>Revisa lo siguiente.</b>

Antes de continuar con este ejemplo y para completar tus aprendizajes del uso de etiquetas, palabras reservadas, expresiones, declaraciones, tipos y operaciones entre ellos, revisa las siguientes notebooks: 
* [<a href="./T01_Etiquetas_y_Palabras_Reservadas.ipynb">etiquetas y palabras reservadas</a>]
* [<a href="./T02_Expr_Decla_Tipos_Oper.ipynb">expresiones, declaraciones, tipos y operadores</a>].
</div>

También es posible usar el símbolo $\rho$ como la etiqueta para la densidad:

In [3]:
𝜌 = 1.038

In [4]:
numerador = M**(2/3) * c * 𝜌**(1/3)
print(numerador, type(numerador), id(numerador))

48.790216719661984 <class 'float'> 2931489793712


### Paso 2. Numerador e `import math`
Implementar la segunda parte de la fórmula: $K \pi^2 (4\pi/3)^{2/3}$. 

Como puedes observar, en esta parte de la fórmula, se requiere del valor de $\pi$. Esta es una constante que podemos obtener del módulo `math` de Python. Para ello debemos importar esta biblioteca y lo podemos hacer de las siguientes maneras:

1. Importar toda la biblioteca `math` y entonces siempre debes anteponer el nombre de la biblioteca para usar sus componentes:
```python
import math
print(math.pi)
```
2. Importar toda la biblioteca `math` y definir un nombre para usar sus componentes:
```python
import math as m
print(m.pi)
```

3. Importar solo la componente de `math` que requieres usar:
```python
from math import pi
print(pi)
```

4. Importar solo la componente de `math` que requieres usar y definir un nombre para esa componente:
```python
from math import pi as PI
print(PI)
```
Incluso podrías usar:
```python
from math import pi as π
print(π)
```

En este ejemplo usaremos esta última forma, pero cualquiera de las anteriores es válida. 

In [5]:
from math import pi as π

K = 5.4e-3
denominador = K * π**2 * (4 * π / 3)**(2/3)
print(denominador, type(denominador), id(denominador))

0.13849026450902358 <class 'float'> 2931489793872


### Paso 3. Tercera parte de la fórmula.

Nos falta implementar $\ln \left[ 0.76 \dfrac{T_o - T_w}{T_y - T_w}\right]$. En este caso se requiere de la función logaritmo natural, la cual puede ser obtenida del módulo `math`. También requerimos definir los valores para $T_o$, $T_w$ y $T_y$.

Para el caso cuando el huevo está a temperatura ambiente tenemos $T_o = 20^oC$. Entonces la implementación final sería como sigue:

In [6]:
from math import log as ln

To = 20  # Temperatura ambiente
Tw = 100
Ty = 70

factor = ln(0.76 * (To - Tw) / (Ty - Tw))
print(factor, type(factor), id(factor))

0.7063924073099658 <class 'float'> 2931489796496


### Paso 4. Fórmula completa.

Ya que tenemos todos los ingredientes de la fórmula, podemos hacer el cálculo final:

In [7]:
t = numerador / denominador * factor
print(t, type(t), id(t))

248.86253747844736 <class 'float'> 2931489793456


El resultado nos da un número flotante que representa el tiempo de cocción del huevo en segundos, cuando su temperatura inicial es de $20^oC$. 

Podemos escribir toda la fórmula en una sola línea como sigue

In [8]:
t = M**(2/3) * c * 𝜌**(1/3) / ( K * π**2 * (4 * π / 3)**(2/3) ) * ln(0.76 * (To - Tw) / (Ty - Tw))
t

248.86253747844736

<div class="alert alert-success">
<b>Precedencia de operadores</b>.
    
Observa que en la implementación anterior se ha tomado en cuenta la precedencia de operadores para que el resultado sea correcto. Revisa estas reglas de precedencia en esta <a href="./T02_Expr_Decla_Tipos_Oper.ipynb">notebook</a> y confirma que la implementación anterior es correcta.
</div>

Ya con todos los ingredientes de la fórmula, podemos calcular el tiempo de cocción para un huevo a temperatura inicial de $T_o = 4^oC$.

In [9]:
To = 4 
t = M**(2/3) * c * 𝜌**(1/3) / ( K * π**2 * (4 * π / 3)**(2/3) ) * ln(0.76 * (To - Tw) / (Ty - Tw))
print(t)

313.09454902221637


Observa que el tiempo es mayor que en el caso anterior, lo cual se debe a que en este último caso el huevo está a una temperatura más baja.

---
## **<font color="DodgerBlue">Ejercicio 2. </font>**

<font color="DarkBlue">
Calcular el tiempo para las temperaturas $20^oC$ y $4^oC$ e imprimir el resultado en el siguiente formato usando cadenas:<br> <br>
    
    `El tiempo de cocción óptimo es: 313.1 [s] (5.2 [m])`
 
<br>
    
En este formato estamos poniendo entre corchetes las unidades del tiempo: `[s]` son segundos y `[m]` son minutos. Entonces debes transformar los segundos en minutos también.
</font>

---

Primero volvemos a calcular los dos tiempos para $20^oC$ y $4^oC$:

In [10]:
To = 4
logaritmo = ln(0.76 * (To - Tw) / (Ty - Tw))
t1 = numerador /denominador * logaritmo

To = 20
logaritmo = ln(0.76 * (To - Tw) / (Ty - Tw))
t2 = numerador /denominador * logaritmo

### Solución 1:

In [11]:
print('El tiempo de cocción óptimo es:', t1,' [s] (', t1/60, ' [m])')
print('El tiempo de cocción óptimo es:', t2,' [s] (', t2/60, ' [m])')

El tiempo de cocción óptimo es: 313.09454902221637  [s] ( 5.218242483703606  [m])
El tiempo de cocción óptimo es: 248.86253747844736  [s] ( 4.147708957974123  [m])


Observa que en esta solución se imprimen muchos decimales en los tiempos, por lo que no es la solución óptima.

### Solución 2:

In [12]:
cadena1 = 'El tiempo de cocción óptimo es: '
cadena2 = str(t1) + ' [s] (' + str(t1/60) + ' [m])'
cadena3 = str(t2) + ' [s] (' + str(t2/60) + ' [m])'
print(cadena1 + cadena2)
print(cadena1 + cadena3)

El tiempo de cocción óptimo es: 313.09454902221637 [s] (5.218242483703606 [m])
El tiempo de cocción óptimo es: 248.86253747844736 [s] (4.147708957974123 [m])


Observa que en esta solución tenemos el mismo problema que en la Solución 1.

### Solución 3:

In [13]:
cadena2 = '{} [s] ( {} [m])'.format(t1, t1/60)
cadena3 = '{} [s] ( {} [m])'.format(t2, t2/60)
print(cadena1 + cadena2)
print(cadena1 + cadena3)

El tiempo de cocción óptimo es: 313.09454902221637 [s] ( 5.218242483703606 [m])
El tiempo de cocción óptimo es: 248.86253747844736 [s] ( 4.147708957974123 [m])


Esta solución sigue presentando el problema de las dos soluciones anteriores.

### Solución 4:

In [16]:
cadena2 = '{:.1f} [s] ({:.1f} [m])'.format(t1, t1/60)
cadena3 = '{:.1f} [s] ({:.1f} [m])'.format(t2, t2/60)
print(cadena1 + cadena2)
print(cadena1 + cadena3)

El tiempo de cocción óptimo es: 313.1 [s] (5.2 [m])
El tiempo de cocción óptimo es: 248.9 [s] (4.1 [m])


Gracias al uso de `{:.1f}` esta solución imprime lo que se solicita correctamente.

### Solución 5:

In [15]:
# Solución 5:
print('El tiempo de cocción óptimo es: {:0.1f} [s] ({:0.1f} [m])'.format(t1, t1/60))
print('El tiempo de cocción óptimo es: {:0.1f} [s] ({:0.1f} [m])'.format(t2, t2/60))

El tiempo de cocción óptimo es: 313.1 [s] (5.2 [m])
El tiempo de cocción óptimo es: 248.9 [s] (4.1 [m])


Esta última solución genera el resultado correcto y es la más recomendada.

<div class="alert alert-success">
<b>Salida estándar</b>.
    
Para conocer más acerca del formato de salida en Python, debes revisar la siguiente  <a href="./T05_Entrada_Salida_Archivos.ipynb">notebook</a>.
</div>

---
## **<font color="DodgerBlue">Ejercicio 3. </font>**

<font color="DarkBlue">
Hacer una lista de tiempos de cocción para temperaturas de huevos, desde la que se tiene en el refrigerador, $4^oC$, hasta temperatura ambiente, $20^oC$, en pasos de $1^oC$.
    
1. Imprimir dos columnas: temperatura inicial ($T_o$) y tiempo ($t$).
2. Imprimir los tiempos en el formato: `min:seg` (por ejemplo 5:30).
3. Imprimir un encabezado para identificar las columnas.

</font>

```
Temperatura 	 Tiempo
 4 		 5:13
 5 		 5:09
...		...
 19 		 4:13
 20 		 4:08
```
---

<div class="alert alert-success">
<b>Estructuras de datos y control de flujo</b>.
    
Para realizar este ejemplo, necesitas conocer las estructuras de datos que pueden usarse en Python, así como las herramientas para el control del flujo del programa. Entonces debes revisar primero las siguientes dos notebooks:
    
<ul>
<li><a href="./T03_Estructura_de_Datos.ipynb">Estructura de datos</a></li>
<li><a href="./T04_Control_de_flujo.ipynb">Control de flujo</a></li>
</ul>

</div>

### Solución.

Recordemos la fórmula $$t = \dfrac{M^{2/3} c \rho^{1/3}}{K \pi^2 (4\pi/3)^{2/3}} \ln \left[ 0.76 \dfrac{T_o - T_w}{T_y - T_w}\right]$$

`numerador` = $M^{2/3} c \rho^{1/3}$

`denominador` = $K \pi^2 (4\pi/3)^{2/3}$

`factor` = $ln \left[ 0.76 \dfrac{T_o - T_w}{T_y - T_w}\right]$

Observamos que lo único que cambia es `factor` cuando cambia $T_o$.

Construyamos primero un ciclo para recorrer los valores de $T_o$ desde $1$ hasta $20$.

In [17]:
for To in range(4,21): # El ciclo va de 4 a 20
    print(To)

4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20


Ahora, dentro del ciclo vamos a implementar la fórmula para calcular el tiempo y se imprime el tiempo y la temperatura inicial $T_o$:

In [18]:
for To in range(4, 21):
    factor = ln(0.76 * (To - Tw) / (Ty - Tw))
    t = numerador / denominador * factor
    print(To, t)

4 313.09454902221637
5 309.4055027800624
6 305.67741828677
7 301.9094604759048
8 298.100767196758
9 294.2504480302776
10 290.3575830395756
11 286.42122145062837
12 282.44038025843554
13 278.4140427535251
14 274.34115696328047
15 270.22063400211084
16 266.05134632399177
17 261.83212587036
18 257.5617621057524
19 253.23899993292162
20 248.86253747844736


Necesitamos convertir el tiempo en el formato `min:seg` :

In [20]:
print('Temperatura \t Tiempo') # Se imprime el encabezado antes del ciclo

for To in range(4, 21):
    factor = ln(0.76 * (To - Tw) / (Ty - Tw))
    t = numerador / denominador * factor
    
    # Conversión del tiempo al formato requerido
    t_min = int(t / 60)
    t_seg = int(t - t_min * 60) 
    
    # Impresión del resultado
    print(' {} \t\t {}:{:02d}'.format(To, t_min, t_seg))

Temperatura 	 Tiempo
 4 		 5:13
 5 		 5:09
 6 		 5:05
 7 		 5:01
 8 		 4:58
 9 		 4:54
 10 		 4:50
 11 		 4:46
 12 		 4:42
 13 		 4:38
 14 		 4:34
 15 		 4:30
 16 		 4:26
 17 		 4:21
 18 		 4:17
 19 		 4:13
 20 		 4:08


Observa que en el formato de impresión de los resultados se usa `{:02d}` lo cual indica que se imprimen 2 dígitos enteros, y cuando solo se tiene un dígito se completa con un cero a la izquierda.

---
## **<font color="DodgerBlue">Ejercicio 4. </font>**

<font color="DarkBlue">
Realizar lo mismo que en el Ejercicio 3 pero usando dos listas: una para las temperaturas y otra para los tiempos.
</font>

---

In [21]:
print('Temperatura \t Tiempo') # Se imprime el encabezado antes del ciclo

# Se definen dos lista vacías
Ts = [] # Lista de temperaturas
ts = [] # Lista de tiempos

for To in range(4,21):
    Ts.append(To) # Se agrega la temperatura a la lista
    
    # Se calcula el tiempo de cocción
    factor = ln(0.76 * (To - Tw) / (Ty - Tw))
    t = numerador / denominador * factor
    
    # Se convierte el tiempo al formato min:seg
    t_min = int(t / 60)
    t_seg = int(t - t_min * 60) 
    
    # Se agrega el tiempo a la lista
    ts.append('{}:{:02d}'.format(t_min, t_seg)) 

# Se imprime el resultado final
for T, t in zip(Ts, ts):
    print(' {} \t\t {}'.format(T, t))

Temperatura 	 Tiempo
 4 		 5:13
 5 		 5:09
 6 		 5:05
 7 		 5:01
 8 		 4:58
 9 		 4:54
 10 		 4:50
 11 		 4:46
 12 		 4:42
 13 		 4:38
 14 		 4:34
 15 		 4:30
 16 		 4:26
 17 		 4:21
 18 		 4:17
 19 		 4:13
 20 		 4:08


---
## **<font color="DodgerBlue">Ejercicio 5. </font>**

<font color="DarkBlue">
Realizar lo mismo que en el Ejercicio 3 pero con un diccionario.
</font>

---

In [22]:
print('Temperatura \t Tiempo') # Se imprime el encabezado antes del ciclo

# Se construye un diccionario vacío
tiempos_huevo = {}

for To in range(4,21):
    
    # Se calcula el tiempo de cocción
    factor = ln(0.76 * (To - Tw) / (Ty - Tw))
    t = numerador / denominador * factor
    
    # Se convierte el tiempo al formato min:seg
    t_min = int(t / 60)
    t_seg = int(t - t_min * 60) 
    
    # Se agrega la temperatura como Key y 
    # el tiempo como value al diccionario:
    tiempos_huevo[To] = '{}:{:02d}'.format(t_min, t_seg)

# Se imprime el resultado final
for key in tiempos_huevo:
    print(' {} \t\t {}'.format(key, tiempos_huevo[key]))

Temperatura 	 Tiempo
 4 		 5:13
 5 		 5:09
 6 		 5:05
 7 		 5:01
 8 		 4:58
 9 		 4:54
 10 		 4:50
 11 		 4:46
 12 		 4:42
 13 		 4:38
 14 		 4:34
 15 		 4:30
 16 		 4:26
 17 		 4:21
 18 		 4:17
 19 		 4:13
 20 		 4:08


Un diccionario se puede recorrer de varias maneras. Observa los siguiente ejemplos:

In [23]:
# Recorrido usando los items, cada item es una tupla
# que contiene el Key y el Value.
for item in tiempos_huevo.items():
    print(item)

(4, '5:13')
(5, '5:09')
(6, '5:05')
(7, '5:01')
(8, '4:58')
(9, '4:54')
(10, '4:50')
(11, '4:46')
(12, '4:42')
(13, '4:38')
(14, '4:34')
(15, '4:30')
(16, '4:26')
(17, '4:21')
(18, '4:17')
(19, '4:13')
(20, '4:08')


In [24]:
# Recorrido usando los items y extrayendo el Key  
# y el Value en las etiquetas k e i
for k, i in tiempos_huevo.items():
    print(k, i)

4 5:13
5 5:09
6 5:05
7 5:01
8 4:58
9 4:54
10 4:50
11 4:46
12 4:42
13 4:38
14 4:34
15 4:30
16 4:26
17 4:21
18 4:17
19 4:13
20 4:08


In [25]:
# Recorrido usando las Keys:
for key in tiempos_huevo.keys():
    print(key)

4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20


In [26]:
# Recorrido unsando los Values:
for val in tiempos_huevo.values():
    print(val)

5:13
5:09
5:05
5:01
4:58
4:54
4:50
4:46
4:42
4:38
4:34
4:30
4:26
4:21
4:17
4:13
4:08


---
## **<font color="DodgerBlue">Ejercicio 6. </font>**

<font color="DarkBlue">
Utilice la versión que más le agrade para escribir las temperaturas y los tiempos de cocción de un huevo duro, y modifíquela para que haga lo siguiente:
    
1. Al principio de la ejecución solicite al usuario:
    - el nombre de un archivo donde va a guardar la tabla de resultados
    - el peso del huevo
2. Imprima la tabla de resultados en pantalla.
3. Guarde la tabla en el archivo.
4. Muestre un mensaje al usuario diciendo el nombre del archivo donde se guardó el resultado.
    
</font>

---

<div class="alert alert-success">
<b>Entrada/salida y gestión de archivos</b>.
    
Para realizar este ejemplo, necesitas saber como manejar la entrada y la salida estándar, así como el manejo de archivos. Para ello debes revisar la siguiente notebook:

<ul>
<li ><a href="./T05_Entrada_Salida_Archivos.ipynb">Entrada, Salida y Archivos</a>.</li>
</ul>
    
</div>

### Solución 1: `dict`

In [27]:
# Se solicita el nombre del archivo
nombre_archivo = input('Nombre del archivo: ')

# Se solicita el peso del huevo.
M = float(input('Peso del huevo = '))

# Imprimimos el encabezado
print('Temperatura \t Tiempo')

# Recalculamos el numerador con el valor de M
numerador = M**(2/3) * c * 𝜌**(1/3)

# Diccionario vacío
tiempos_huevo = {} 

for To in range(4,21):
    
    # Se calcula el tiempo de cocción
    factor = ln(0.76 * (To - Tw) / (Ty - Tw))
    t = numerador / denominador * factor
    
    # Se convierte el tiempo al formato min:seg
    t_min = int(t / 60)
    t_seg = int(t - t_min * 60) 
    
    # Se guarda la información en el diccionario
    tiempos_huevo[To] = '{}:{:02d}'.format(t_min, t_seg)

# Usando un gestor de contexto abrimos el archivo.
with open(nombre_archivo, 'w') as archivo_abierto:
    
    # Escribimos el encabezado en el archivo
    archivo_abierto.write('Temperatura \t Tiempo\n')
    
    # Escribimos los tiempos y temperaturas en el archivo
    for key in tiempos_huevo:
        print(' {} \t\t {}'.format(key, tiempos_huevo[key]))
        archivo_abierto.write(' {} \t\t {}\n'.format(key, tiempos_huevo[key]))

print('La tabla de tiempos de cocción se guardó en el archivo: "{}"'.format(nombre_archivo))

Nombre del archivo:  borrame
Peso del huevo =  47


Temperatura 	 Tiempo
 4 		 5:13
 5 		 5:09
 6 		 5:05
 7 		 5:01
 8 		 4:58
 9 		 4:54
 10 		 4:50
 11 		 4:46
 12 		 4:42
 13 		 4:38
 14 		 4:34
 15 		 4:30
 16 		 4:26
 17 		 4:21
 18 		 4:17
 19 		 4:13
 20 		 4:08
La tabla de tiempos de cocción se guardó en el archivo: "borrame"


Verifica que el archivo se guardó correctamente y verificar su contenido.

### Solución 2: `list`

In [28]:
# Se solicita el nombre del archivo
nombre_archivo = input('Nombre del archivo: ')

# Se solicita el peso del huevo.
M = float(input('Peso del huevo = '))

# Imprimimos el encabezado
print('Temperatura \t Tiempo')

# Recalculamos el numerador con el valor de M
numerador = M**(2/3) * c * 𝜌**(1/3)

# Creamos dos listas vacías
Ts = []
ts = []

for To in range(4,21):
    Ts.append(To)
    
    # Se calcula el tiempo de cocción 
    factor = ln(0.76 * (To - Tw) / (Ty - Tw))
    t = numerador / denominador * factor

    # Se convierte el tiempo al formato min:seg
    t_min = int(t / 60)
    t_seg = int(t - t_min * 60) 

    # Se agrega el tiempo a la lista
    t1 = '{}:{:02d}'.format(t_min, t_seg)
    ts.append(t1)

# Usando un gestor de contexto abrimos el archivo.
with open(nombre_archivo, 'w') as archivo_abierto:
    # Escribimos el encabezado en el archivo
    archivo_abierto.write('Temperatura \t Tiempo\n')
    
    # Escribimos los tiempos y temperaturas en el archivo    
    for i in range(len(Ts)):
        print(' {} \t\t {}'.format(Ts[i], ts[i]))
        archivo_abierto.write(' {} \t\t {}\n'.format(Ts[i], ts[i]))

print('La tabla de tiempos de cocción se guardó en el archivo: "{}"'.format(nombre_archivo))

Nombre del archivo:  borrame
Peso del huevo =  47


Temperatura 	 Tiempo
 4 		 5:13
 5 		 5:09
 6 		 5:05
 7 		 5:01
 8 		 4:58
 9 		 4:54
 10 		 4:50
 11 		 4:46
 12 		 4:42
 13 		 4:38
 14 		 4:34
 15 		 4:30
 16 		 4:26
 17 		 4:21
 18 		 4:17
 19 		 4:13
 20 		 4:08
La tabla de tiempos de cocción se guardó en el archivo: "borrame"


Verifica que el archivo se guardó correctamente y verificar su contenido.

---
## **<font color="DodgerBlue">Ejercicio 7. </font>**

<font color="DarkBlue">
Modificar el código del ejercicio 4 para que el nombre del archivo contenga el peso del huevo. Por ejemplo, si el usuario teclea <b>tiempo_coccion</b> en el nombre del archivo y <b>67</b> en el peso del huevo, el resultado final se almacenará en un archivo de nombre <b>tiempo_coccion_67</b>.
</font>

---

In [29]:
# Se solicita el nombre del archivo
nombre_archivo = input('Nombre del archivo: ')

# Se solicita el peso del huevo.
M = float(input('Peso del huevo = '))

# Imprimimos el encabezado
print('Temperatura \t Tiempo')

# Recalculamos el numerador con el valor de M
numerador = M**(2/3) * c * 𝜌**(1/3)

# Construcción del nombre del archivo
etiqueta = str(int(M))
nombre_archivo += '_' + etiqueta

# Construimos un diccionario vacío
tiempos_huevo = {} 

for To in range(4,21):
    
    # Se calcula el tiempo de cocción 
    factor = ln(0.76 * (To - Tw) / (Ty - Tw))
    t = numerador / denominador * factor
    
    # Se convierte el tiempo al formato min:seg
    t_min = int(t / 60)
    t_seg = int(t - t_min * 60) 
    
    # Se guarda la información en el diccionario
    tiempos_huevo[To] = '{}:{:02d}'.format(t_min, t_seg)

# Usando un gestor de contexto abrimos el archivo.
with open(nombre_archivo, 'w') as archivo_abierto:
    
    # Escribimos el encabezado en el archivo    
    archivo_abierto.write('Temperatura \t Tiempo\n')

    # Escribimos los tiempos y temperaturas en el archivo    
    for key in tiempos_huevo:
        print(' {} \t\t {}'.format(key, tiempos_huevo[key]))
        archivo_abierto.write(' {} \t\t {}\n'.format(key, tiempos_huevo[key]))
    
print('La tabla de tiempos de cocción se guardó en el archivo: "{}"'.format(nombre_archivo))

Nombre del archivo:  borrame
Peso del huevo =  47


Temperatura 	 Tiempo
 4 		 5:13
 5 		 5:09
 6 		 5:05
 7 		 5:01
 8 		 4:58
 9 		 4:54
 10 		 4:50
 11 		 4:46
 12 		 4:42
 13 		 4:38
 14 		 4:34
 15 		 4:30
 16 		 4:26
 17 		 4:21
 18 		 4:17
 19 		 4:13
 20 		 4:08
La tabla de tiempos de cocción se guardó en el archivo: "borrame_47"


---
## **<font color="DodgerBlue">Ejercicio 8. </font>**

<font color="DarkBlue">
Determina el intervalo de temperaturas, en que debería estar la temperatura del huevo inicialmente para que el tiempo de cocción sea exactamente de 6 minutos para un huevo de 67 gramos.
</font>

---

In [30]:
# Peso del huevo
M = 67

# Recalculamos el numerador con el valor de M
numerador = M**(2/3) * c * 𝜌**(1/3)

To = 1

count = 0
while True:
    
    # Se calcula el tiempo de cocción 
    factor = ln(0.76 * (To - Tw) / (Ty - Tw))
    t = numerador / denominador * factor

    # Se convierte el tiempo al formato min:seg
    t_min = int(t / 60)
    t_seg = int(t - t_min * 60) 
    
    print('{}:{:02d} \t {}'.format(t_min, t_seg, To))
    
    if t_min > 6:
        To += 0.5
    elif t_min < 6:
        To -= 0.5
    else:
        if t_seg > 0:
            To += 0.5
        else:
            print('La temperatura inicial {} [C] requiere de {}:{:02d}'.format(To, t_min, t_seg))
            break
            
    count += 1 
    if count > 100:
        break

6:50 	 1
6:48 	 1.5
6:45 	 2.0
6:43 	 2.5
6:41 	 3.0
6:38 	 3.5
6:36 	 4.0
6:34 	 4.5
6:31 	 5.0
6:29 	 5.5
6:27 	 6.0
6:24 	 6.5
6:22 	 7.0
6:20 	 7.5
6:17 	 8.0
6:15 	 8.5
6:12 	 9.0
6:10 	 9.5
6:07 	 10.0
6:05 	 10.5
6:02 	 11.0
6:00 	 11.5
La temperatura inicial 11.5 [C] requiere de 6:00


---
## **<font color="DodgerBlue">Ejercicio 9. </font>**

<font color="DarkBlue">
Construye la función <b>calc_tiempo()</b> para calcular el tiempo de cocción. Esta función debe recibir todos los parámetros necesarios para evaluar la fórmula con valores por omisión, excepto para $M$ y $T_o$. La función debe regresar el tiempo calculado en una cadena en formato <b>min:seg</b>.
</font>

---

<div class="alert alert-success">
<b>Funciones y manejo de excepciones</b>.
    
Para realizar los ejemplos 9 y 10, necesitas saber como construir y usar funciones, documentar dichas funciones y manejar excepciones. Para ello debes revisar las siguientes notebooks:

<ul>
<li><a href="./T06_Funciones_y_Documentacion.ipynb">Funciones y documentación</a>.</li>
<li><a href="./T07_Excepciones.ipynb">Manejo de excepciones</a>.</li>
</ul>
    
</div>

In [32]:
from math import pi as π
from math import log as ln

def calc_tiempo(M, To, c = 3.7, 𝜌 = 1.038, K = 5.4e-3, Tw = 100, Ty = 70):

    # Se calcula el tiempo de cocción 
    t = M**(2/3) * c * 𝜌**(1/3) / ( K * π**2 * (4 * π / 3)**(2/3) ) * ln(0.76 * (To - Tw) / (Ty - Tw))

    # Se convierte el tiempo al formato min:seg
    t_min = int(t / 60)
    t_seg = int(t - t_min * 60)
    
    return '{}:{:02d}'.format(t_min, t_seg)

In [33]:
calc_tiempo(47, 4)

'5:13'

---
## **<font color="DodgerBlue">Ejercicio 10. </font>**

<font color="DarkBlue">
Usando la función construida en el ejemplo 9, modfíquela para que maneje excepciones. En este caso, la excepción debe manejar el error de tipo <b>TypeError</b>, es decir cuando el usuario use un tipo de dato incorrecto, por ejemplo: <b>calc_tiempo('47', 4)</b>. Documente la función usando <i>docstring</i>. 
</font>

---

In [34]:
import math

def calc_tiempo(M, To, c = 3.7, 𝜌 = 1.038, K = 5.4e-3, Tw = 100, Ty = 70):
    """
    Esta función calcula el tiempo de un huevo dados ciertos parámetros físicos.
    
    Parameters
    ----------
    M : float
    Masa del huevo ek gramos.
    
    To : float
    Temperatura original del huevo, en grados Celsius, antes de meterlo al agua.
    
    c : float
    Capacidad calorífica específica.
    
    𝜌 : float
    Densidad.
    
    K : float
    Conductividad térmica.
    
    Tw : float
    Temperatura de ebullición del agua.
    
    Ty : float
    Temperatura que debe alcanzar la yema.

    Returns
    -------
    str: Tiempo de cocción del huevo en format min:seg
    """
    π = math.pi
    
    ocurre_error = False
    try:
        # Se calcula el tiempo de cocción 
        t = M**(2/3) * c * 𝜌**(1/3) / ( K * π**2 * (4 * π / 3)**(2/3) ) * ln(0.76 * (To - Tw) / (Ty - Tw))

        # Se convierte el tiempo al formato min:seg
        t_min = int(t / 60)
        t_seg = int(t - t_min * 60)
        
        resultado = '{}:{:02d}'.format(t_min, t_seg)
    except TypeError as detalles:
        ocurre_error = True
        resultado = "Ocurrió un error (TypeError): \n {}".format(detalles)
    except:
        ocurre_error = True
        resultado = "Ocurrió algo misterioso"
        
    if ocurre_error:
        print(resultado)
        print('Revisa el correcto funcionamiento de la función.')
    else:
        return resultado    


In [37]:
calc_tiempo(47,4)

'5:13'

Debido a que esta última función ya está documentada con *docstring*, es posible poner el cursor en el nombre de la función y teclear [Shift+Tab] para obtener la ayuda.

**Si deseas probar tus aprendizajes, intenta realizar e siguiente miniproyecto.**

---
## **<font color="DodgerBlue">Miniproyecto 1. </font>** (Opcional)

<font color="DarkBlue">Calcule las tablas de tiempos para huevos con peso mínimo de 47 y máximo de 67 gramos, en pasos de 1 gramo, para las temperaturas iniciales desde $T_o = 4^o C$ y hasta $T_o = 20^o C$ en pasos de $1^o C$. Imprima al principio de cada tabla la leyenda: `Peso del huevo = 67 [g]`. Almacenar estas tablas en una lista de diccionarios, donde cada elemento de la lista es un diccionario que contiene la información para una determinada $M$. Guardar las tablas de manera secuencial en un archivo cuyo nombre debe ser proporcionado por el usuario. Finalmente preguntar al usuario la tabla de tiempos que desea ver (con base en el peso del huevo) en pantalla. </font>

---

In [94]:
### BEGIN SOLUTION
# Se solicita el nombre del archivo
nombre_archivo = input('Nombre del archivo: ')

# Recalculamos el numerador con el valor de M
numerador = M**(2/3) * c * 𝜌**(1/3)

# Construimos una lista vacía
lista_de_diccionarios = []

with open(nombre_archivo, 'w') as archivo_abierto:
    for M in range(47, 68):
        dummy_dict = {}

        print('Peso del huevo = {} [g]'.format(M))
        print('Temperatura \t Tiempo')
        archivo_abierto.write('Peso del huevo = {} [g]\n'.format(M))
        archivo_abierto.write('Temperatura \t Tiempo\n')
        
        for To in range(4,21):
            t = calc_tiempo(M,To)
            dummy_dict[To] = t
            print(' {} \t\t {}'.format(To, t))
            archivo_abierto.write(' {} \t\t {}\n'.format(To, t))

        lista_de_diccionarios.append(dummy_dict)
    
print('La tabla de tiempos de cocción se guardó en el archivo: "{}"'.format(nombre_archivo))
### END SOLUTION

Nombre del archivo:  borrame


Peso del huevo = 47 [g]
Temperatura 	 Tiempo
 4 		 5:13
 5 		 5:09
 6 		 5:05
 7 		 5:01
 8 		 4:58
 9 		 4:54
 10 		 4:50
 11 		 4:46
 12 		 4:42
 13 		 4:38
 14 		 4:34
 15 		 4:30
 16 		 4:26
 17 		 4:21
 18 		 4:17
 19 		 4:13
 20 		 4:08
Peso del huevo = 48 [g]
Temperatura 	 Tiempo
 4 		 5:17
 5 		 5:13
 6 		 5:09
 7 		 5:06
 8 		 5:02
 9 		 4:58
 10 		 4:54
 11 		 4:50
 12 		 4:46
 13 		 4:42
 14 		 4:38
 15 		 4:34
 16 		 4:29
 17 		 4:25
 18 		 4:21
 19 		 4:16
 20 		 4:12
Peso del huevo = 49 [g]
Temperatura 	 Tiempo
 4 		 5:21
 5 		 5:18
 6 		 5:14
 7 		 5:10
 8 		 5:06
 9 		 5:02
 10 		 4:58
 11 		 4:54
 12 		 4:50
 13 		 4:46
 14 		 4:42
 15 		 4:37
 16 		 4:33
 17 		 4:29
 18 		 4:24
 19 		 4:20
 20 		 4:15
Peso del huevo = 50 [g]
Temperatura 	 Tiempo
 4 		 5:26
 5 		 5:22
 6 		 5:18
 7 		 5:14
 8 		 5:10
 9 		 5:06
 10 		 5:02
 11 		 4:58
 12 		 4:54
 13 		 4:50
 14 		 4:45
 15 		 4:41
 16 		 4:37
 17 		 4:32
 18 		 4:28
 19 		 4:23
 20 		 4:19
Peso del huevo = 51 [g]
Temp

<div class="alert alert-success">
<b>Revisa lo siguiente.</b>

Antes de continuar con los siguientes ejemplos es importante que revises las siguientes notebooks: 
    
* [<a href="./T08_IterablesMapFilter.ipynb">T08_IterablesMapFilter.ipynb</a>]
* [<a href="./T09_LambdaExpressions_Reduce.ipynb">T09_LambdaExpressions_Reduce.ipynb</a>].
* [<a href="./T10_Comprehensions.ipynb">T10_Comprehensions.ipynb</a>]
* [<a href="./T11_IteradoresGeneradores.ipynb">T11_IteradoresGeneradores.ipynb</a>].
* [<a href="./T12_Decoradores.ipynb">T12_Decoradores.ipynb</a>]
* [<a href="./T13_BibliotecaEstandar.ipynb">T13_BibliotecaEstandar.ipynb</a>].
</div>

---
## **<font color="DodgerBlue">Ejercicio 11. </font>**

<font color="DarkBlue">
Construir una lista de temperaturas de $4^oC$ hasta $20^oC$ solo almacenando los valores pares usando una <i>comprehension list</i> y aplicar la función <b>calc_tiempo()</b> para generar una lista de los tiempos correspondientes e imprimir las columas de temperaturas y tiempos como antes.
</font>

---

In [39]:
T_pares = [T for T in range(4,21,2)]
T_pares

[4, 6, 8, 10, 12, 14, 16, 18, 20]

In [44]:
t_pares = [calc_tiempo(47, T) for T in T_pares]
t_pares

['5:13', '5:05', '4:58', '4:50', '4:42', '4:34', '4:26', '4:17', '4:08']

In [45]:
[calc_tiempo(47, T) for T in range(4,21,2)]

['5:13', '5:05', '4:58', '4:50', '4:42', '4:34', '4:26', '4:17', '4:08']

In [48]:
T_pares = [T for T in range(4,21,2)]
t_pares = [calc_tiempo(47, T) for T in T_pares]

print('Temperatura \t Tiempo')

for T, t in zip(T_pares, t_pares):
    print(' {} \t\t {}'.format(T, t))

Temperatura 	 Tiempo
 4 		 5:13
 6 		 5:05
 8 		 4:58
 10 		 4:50
 12 		 4:42
 14 		 4:34
 16 		 4:26
 18 		 4:17
 20 		 4:08


---
## **<font color="DodgerBlue">Ejercicio 12. </font>**

<font color="DarkBlue">
Convertir la lista anterior a grados Fahrenheit usando un mapeo y una función anónima para la conversión. Imprimir el resultado como antes pero usando los grados Fahrenheit.
</font>

---

In [49]:
toFahrenheit = lambda T: (9/5)*T + 32

In [65]:
T_pares = [T for T in range(4,21,2)]
F_pares = map(toFahrenheit, T_pares)
F_pares

In [66]:
list(F_pares)

[39.2, 42.8, 46.4, 50.0, 53.6, 57.2, 60.8, 64.4, 68.0]

In [70]:
t_pares = [calc_tiempo(47, T) for T in T_pares]
F_pares = list(map(toFahrenheit, T_pares))

print('Temperatura \t Tiempo')

for T, t in zip(list(F_pares), t_pares):
    print(' {} \t\t {}'.format(T, t))

Temperatura 	 Tiempo
 39.2 		 5:13
 42.8 		 5:05
 46.4 		 4:58
 50.0 		 4:50
 53.6 		 4:42
 57.2 		 4:34
 60.8 		 4:26
 64.4 		 4:17
 68.0 		 4:08


---
## **<font color="DodgerBlue">Ejercicio 13. </font>**

<font color="DarkBlue">
Crear una función anónima que detecte un número par y usarla en conjunción con la función <b>filter()</b> para generar la lista de temperaturas pares, luego calcular los tiempos de cocción y finalmente imprimir la lista de temperaturas en grados Fahrenheit junto con los tiempos.
</font>

---

In [81]:
es_par = lambda n: True if n%2 == 0 else False

In [85]:
list(range(4,21))

[4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20]

In [86]:
filter(es_par, list(range(4,21)))

In [87]:
list(filter(es_par, list(range(4,21))))

[4, 6, 8, 10, 12, 14, 16, 18, 20]

In [88]:
T_pares = list(filter(es_par, list(range(4,21))))
t_pares = [calc_tiempo(47, T) for T in T_pares]
F_pares = list(map(toFahrenheit, T_pares))

print('Temperatura \t Tiempo')

for T, t in zip(F_pares, t_pares):
    print(' {} \t\t {}'.format(T, t))

Temperatura 	 Tiempo
 39.2 		 5:13
 42.8 		 5:05
 46.4 		 4:58
 50.0 		 4:50
 53.6 		 4:42
 57.2 		 4:34
 60.8 		 4:26
 64.4 		 4:17
 68.0 		 4:08


---
## **<font color="DodgerBlue">Ejercicio 14. </font>**

<font color="DarkBlue">
Realiza lo mismo que en el Ejercicio 13 pero para número impares.
</font>

---

In [93]:
### BEGIN SOLUTION
es_impar = lambda n: True if n%2 else False

T_impares = list(filter(es_impar, list(range(4,21))))
t_impares = [calc_tiempo(47, T) for T in T_impares]
F_impares = list(map(toFahrenheit, T_impares))

print('Temperatura \t Tiempo')

for T, t in zip(F_impares, t_impares):
    print(' {} \t\t {}'.format(T, t))
### END SOLUTION

Temperatura 	 Tiempo
 41.0 		 5:09
 44.6 		 5:01
 48.2 		 4:54
 51.8 		 4:46
 55.400000000000006 		 4:38
 59.0 		 4:30
 62.6 		 4:21
 66.2 		 4:13


---
## **<font color="DodgerBlue">Miniproyecto 2. </font>** (Opcional)

<font color="DarkBlue">Modifica el miniproyecto 1 para usar <b>lambda</b>, <b>map</b>, <b>filter</b> y que además pregunte al usuario si desea hacer el cálculo para temperaturas pares o impares. El resultado debe estar en grados Fahrenheit. </font>

---

In [103]:
### BEGIN SOLUTION
# Se solicita el nombre del archivo
nombre_archivo = input('Nombre del archivo: ')
par_o_impar = input('¿Pares [p] o Impares [i] ?')

toFahrenheit = lambda T: (9/5)*T + 32

if par_o_impar.lower() == 'P'.lower():
    es_par = lambda n: True if n%2 == 0 else False
    Ts = list(filter(es_par, list(range(4,21))))
elif par_o_impar.lower() == 'I'.lower():
    es_impar = lambda n: True if n%2 else False
    Ts = list(filter(es_impar, list(range(4,21))))

with open(nombre_archivo, 'w') as archivo_abierto:
    for M in range(47, 68):
        ts = [calc_tiempo(M, T) for T in Ts]
        Fs = list(map(toFahrenheit, Ts))        
        
        dummy_dict = {}

        print('Peso del huevo = {} [g]'.format(M))
        print('Temperatura \t Tiempo')
        archivo_abierto.write('Peso del huevo = {} [g]\n'.format(M))
        archivo_abierto.write('Temperatura \t Tiempo\n')
        
        for F, t in zip(Fs, ts):
            dummy_dict[F] = t
            print(' {} \t\t {}'.format(F, t))
            archivo_abierto.write(' {} \t\t {}\n'.format(F, t))

        lista_de_diccionarios.append(dummy_dict)
    
print('La tabla de tiempos de cocción se guardó en el archivo: "{}"'.format(nombre_archivo))
### END SOLUTION

Nombre del archivo:  borrame
¿Pares [p] o Impares [i] ? I


Peso del huevo = 47 [g]
Temperatura 	 Tiempo
 41.0 		 5:09
 44.6 		 5:01
 48.2 		 4:54
 51.8 		 4:46
 55.400000000000006 		 4:38
 59.0 		 4:30
 62.6 		 4:21
 66.2 		 4:13
Peso del huevo = 48 [g]
Temperatura 	 Tiempo
 41.0 		 5:13
 44.6 		 5:06
 48.2 		 4:58
 51.8 		 4:50
 55.400000000000006 		 4:42
 59.0 		 4:34
 62.6 		 4:25
 66.2 		 4:16
Peso del huevo = 49 [g]
Temperatura 	 Tiempo
 41.0 		 5:18
 44.6 		 5:10
 48.2 		 5:02
 51.8 		 4:54
 55.400000000000006 		 4:46
 59.0 		 4:37
 62.6 		 4:29
 66.2 		 4:20
Peso del huevo = 50 [g]
Temperatura 	 Tiempo
 41.0 		 5:22
 44.6 		 5:14
 48.2 		 5:06
 51.8 		 4:58
 55.400000000000006 		 4:50
 59.0 		 4:41
 62.6 		 4:32
 66.2 		 4:23
Peso del huevo = 51 [g]
Temperatura 	 Tiempo
 41.0 		 5:26
 44.6 		 5:18
 48.2 		 5:10
 51.8 		 5:02
 55.400000000000006 		 4:53
 59.0 		 4:45
 62.6 		 4:36
 66.2 		 4:27
Peso del huevo = 52 [g]
Temperatura 	 Tiempo
 41.0 		 5:30
 44.6 		 5:22
 48.2 		 5:14
 51.8 		 5:06
 55.400000000000006 		 4:57
 59.0 		 4:49
 6